src/relations/mysql_provider.py

```diff
@@ -309,7 +315,38 @@ def _on_database_broken(self, event: RelationBrokenEvent) -> None:
         relation_id = event.relation.id
         try:
             self.charm._mysql.delete_users_for_relation(relation_id)
-            logger.info(f"Removed user for relation {relation_id}")
+            logger.info(f"Removed user(s) for relation {relation_id}")
         except MySQLDeleteUsersForRelationError:
-            logger.error(f"Failed to delete user for relation {relation_id}")
+            logger.error(f"Failed to delete user(s) for relation {relation_id}")
+            return
+
+    def _on_database_provides_relation_departed(self, event: RelationDepartedEvent) -> None:
+        """Remove MySQL Router cluster metadata & user for departing unit."""
+        if not self.charm.unit.is_leader():
+            return
+        if event.departing_unit.app.name == self.charm.app.name:
             return
+
+        users = self.charm._mysql.get_mysql_router_users_for_unit(
+            relation_id=event.relation.id, mysql_router_unit_name=event.departing_unit.name
+        )
+        if not users:
+            return
+
+        if len(users) > 1:
+            logger.error(
+                f"More than one router user for departing unit {event.departing_unit.name}"
+            )
+            return
+
+        user = users[0]
+        try:
+            self.charm._mysql.delete_user(user.username)
+            logger.info(f"Deleted router user {user.username}")
+        except MySQLDeleteUserError:
+            logger.error(f"Failed to delete user {user.username}")
```

Comment:
maybe we should defer/retry if the deletion fails.
up to @paulomach

Reply:
I'd agree - if there's a error dropping a user it maybe the case to retry - I've also commented on the lib method to make it harder to fail.

Reply:
Sure